[metricbeat] [gcp] add billing additional fields

gpop63 opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
- Cleanup
- Docs
-->

## Overview

Added functionality to check if a billing data table is detailed, using the prefix `gcp_billing_export_resource_v1_` for identification.

Added a `row` struct in the code to manage billing data fields, replacing the previous method that accessed fields by indexes. This change makes the data handling clearer and reduces the likelihood of errors due to misindexing.

Previously, `max` aggregation displayed daily costs on the billing dashboard for one document per day. With multiple documents now generated daily, `sum` aggregation is needed to accurately add the total cost from all documents.

<b>Table formats:</b>
- Standard cost usage table format: `gcp_billing_export_v1_<BILLING_ACCOUNT_ID>`
- Detailed cost usage table format: `gcp_billing_export_resource_v1_<BILLING_ACCOUNT_ID>`

<b>Common fields:</b> Additional fields applicable to both standard and detailed cost data:
- `sku.id`
- `sku.description`
- `service.id`
- `service.description`
- `tags`

<b>Detailed fields:</b>
- `price.effective_price`

<b>Documentation:</b>
- [Standard Usage Cost Data](https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/standard-usage)
- [Detailed Usage Cost Data](https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/detailed-usage)

<!-- Mandatory
Explain here the changes you made on the PR.

Please explain:

- WHAT: patterns used, algorithms implemented, design architecture, message processing, etc.
- WHY:  the rationale/motivation for the changes

This text will be pasted into the squash dialog when the change is committed and will be
a long term historical record of the change to help future contributors understand the
change, please help them by making it clear and comprehensive, they may be you.

If the commit title is adequate to describe both of these things, The text here may be omitted
or replaced with "See title". The title of the PR will be used as the commit message title when
the merge is made and the "See title" marker will be removed if present.

The text here and the PR title will be subject to the PR review process.
-->

## Checklist

<!-- Mandatory
Add a checklist of things that are required to be reviewed in order to have the PR approved

List here all the items you have verified BEFORE sending this PR. Please DO NOT remove any item, striking through those that do not apply. (Just in case, strikethrough uses two tildes. ~~Scratch this.~~)
-->

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

<!-- Recommended
Add a checklist of things that are required to be reviewed in order to have the PR approved
-->
- [ ]

## How to test this PR locally

<!-- Recommended
Explain here how this PR will be tested by the reviewer: commands, dependencies, steps, etc.
-->

## Related issues

<!-- Recommended
Link related issues below. Insert the issue link or reference after the word "Closes" if merging this should automatically close it.

- Closes #123
- Relates #123
- Requires #123
- Superseds #123
-->
- Closes #36886

## Use cases

<!-- Recommended
Explain here the different behaviors that this PR introduces or modifies in this project, user roles, environment configuration, etc.

If you are familiar with Gherkin test scenarios, we recommend its usage: https://cucumber.io/docs/gherkin/reference/
-->

## Screenshots

<!-- Optional
Add here screenshots about how the project will be changed after the PR is applied. They could be related to web pages, terminal, etc, or any other image you consider important to be shared with the team.
-->

## Logs

<!-- Recommended
Paste here output logs discovered while creating this PR, such as stack traces or integration logs, or any other output you consider important to be shared with the team.
-->

elasticmachine commented:
## :green_heart: Build Succeeded


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36902/detail/PR-36902/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36902/detail/PR-36902/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36902/detail/PR-36902/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2FPR-36902/detail/PR-36902/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_36902.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2023-10-18T22:55:13.064Z&rangeTo=2023-10-18T23:15:13.064Z&transactionName=BUILD+Beats%2Fbeats%2FPR-%7Bnumber%7D&transactionType=job&latencyAggregationType=avg&traceId=d0566e4d840d18a5d90eda5297b431cb&transactionId=dab8f8a4e0b9dc21)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats





* Duration: 52 min 36 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














## :grey_exclamation: Flaky test report
No test was executed to be analysed.



## :robot: GitHub comments

<details><summary>Expand to view the GitHub comments</summary>
<p>

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.


- `/package` : Generate the packages and run the E2E tests.


- `/beats-tester` : Run the installation tests with beats-tester.


- `run` `elasticsearch-ci/docs` : Re-trigger the docs validation. (use unformatted text in the comment!)




</p>
</details>

<!--COMMENT_GENERATED_WITH_ID_comment.id-->

gpop63 commented on x-pack/metricbeat/module/gcp/billing/billing.go:
```diff
@@ -314,27 +381,135 @@ func generateEventID(currentDate string, rowItems []bigquery.Value) string {
+		return createDetailedQuery(tableName, month, costType)
+	}
+
+	return createStandardQuery(tableName, month, costType)
```
The table pattern will always match the standard pattern even if user forgets to set it.

https://github.com/elastic/beats/blob/b7af6f91a169fd9b14e88989de6dcc292235d0b5/x-pack/metricbeat/module/gcp/billing/billing.go#L131-L134

gpop63 commented:
<details>
<summary>Standard document example:</summary>
<p>

```json
{
  "_index": ".ds-metricbeat-8.12.0-2023.10.19-000001",
  "_id": "eh2USYsBN2wa5YEjwRKW",
  "_score": 1,
  "_source": {
    "@timestamp": "2023-10-19T20:16:01.524Z",
    "metricset": {
      "name": "billing",
      "period": 60000
    },
    "service": {
      "type": "gcp"
    },
    "gcp": {
      "billing": {
        "service_description": "Compute Engine",
        "invoice_month": "202310",
        "project_id": "robust-catalyst-399814",
        "billing_account_id": "011702-58A742-BEB4E7",
        "total": 0,
        "sku_id": "0D56-2F80-52A5",
        "cost_type": "regular",
        "service_id": "6F81-5844-456A",
        "sku_description": "Network Inter Region Ingress from Jakarta to Americas",
        "tags": {},
        "project_name": "My Project 70130"
      }
    },
    "cloud.project.id": "robust-catalyst-399814",
    "cloud.project.name": "My Project 70130",
    "host": {
      "os": {
        "platform": "ubuntu",
        "version": "22.04.2 LTS (Jammy Jellyfish)",
        "family": "debian",
        "name": "Ubuntu",
        "kernel": "5.10.102.1-microsoft-standard-WSL2",
        "codename": "jammy",
        "type": "linux"
      },
      "containerized": false,
      "ip": [
        "172.30.189.193",
        "fe80::215:5dff:fecc:d3ab"
      ],
      "mac": [
        "00-15-5D-CC-D3-AB",
        "5A-2B-CE-65-CF-33",
        "62-74-42-62-2C-60"
      ],
      "name": "gab",
      "hostname": "gab",
      "architecture": "x86_64"
    },
    "cloud.account.id": "011702-58A742-BEB4E7",
    "event": {
      "duration": 3895440085,
      "dataset": "gcp.billing",
      "module": "gcp"
    },
    "cloud.provider": "gcp",
    "ecs": {
      "version": "8.0.0"
    },
    "agent": {
      "id": "afdc4ab4-827a-4058-97e7-267618840f1d",
      "name": "gab",
      "type": "metricbeat",
      "version": "8.12.0",
      "ephemeral_id": "a071aaea-4ff5-4ec7-b7b0-672f358d2438"
    }
  }
}
```

</p>
</details>

<details>
<summary>Detailed document example:</summary>
<p>

```json
{
  "_index": ".ds-metricbeat-8.12.0-2023.10.19-000001",
  "_id": "3h2RSYsBN2wa5YEjUBFz",
  "_score": 1,
  "_source": {
    "@timestamp": "2023-10-19T20:12:17.902Z",
    "event": {
      "dataset": "gcp.billing",
      "module": "gcp",
      "duration": 3092063008
    },
    "metricset": {
      "period": 60000,
      "name": "billing"
    },
    "service": {
      "type": "gcp"
    },
    "host": {
      "ip": [
        "172.30.189.193",
        "fe80::215:5dff:fecc:d3ab"
      ],
      "mac": [
        "00-15-5D-CC-D3-AB",
        "5A-2B-CE-65-CF-33",
        "62-74-42-62-2C-60"
      ],
      "name": "gab",
      "hostname": "gab",
      "architecture": "x86_64",
      "os": {
        "name": "Ubuntu",
        "kernel": "5.10.102.1-microsoft-standard-WSL2",
        "codename": "jammy",
        "type": "linux",
        "platform": "ubuntu",
        "version": "22.04.2 LTS (Jammy Jellyfish)",
        "family": "debian"
      },
      "containerized": false
    },
    "gcp": {
      "billing": {
        "effective_price": 0.00292353,
        "tags": {
          "compute.googleapis.com/machine_spec": "e2-medium",
          "compute.googleapis.com/memory": "4096",
          "compute.googleapis.com/cores": "2"
        },
        "billing_account_id": "011702-58A742-BEB4E7",
        "cost_type": "regular",
        "sku_id": "F449-33EC-A5EF",
        "sku_description": "E2 Instance Ram running in Americas",
        "service_description": "Compute Engine",
        "invoice_month": "202310",
        "project_name": "My Project 70130",
        "service_id": "6F81-5844-456A",
        "project_id": "robust-catalyst-399814",
        "total": 1.700642
      }
    },
    "cloud.provider": "gcp",
    "cloud.project.id": "robust-catalyst-399814",
    "cloud.account.id": "011702-58A742-BEB4E7",
    "agent": {
      "ephemeral_id": "dac27eeb-f92c-43be-a0e6-ad3863bc5a27",
      "id": "afdc4ab4-827a-4058-97e7-267618840f1d",
      "name": "gab",
      "type": "metricbeat",
      "version": "8.12.0"
    },
    "ecs": {
      "version": "8.0.0"
    },
    "cloud.project.name": "My Project 70130"
  }
}
```

</p>
</details>

mrg-elastic committed:
GCP Billing Enhancement
Implemented a check to differentiate between standard and detailed cost usage tables
Refactoring the 'row' structure used for storing BigQuery results, optimizing data handling
Added support for tags
Introduced new cost usage data fields

mrg-elastic committed:
add tests
fix older tests

mrg-elastic committed:
add changelog entry

lucian-ioan commented on x-pack/metricbeat/module/gcp/billing/billing.go:
```diff
@@ -315,26 +380,134 @@ func generateEventID(currentDate string, rowItems []bigquery.Value) string {
+	return query
+}
+
+func createDetailedQuery(tableName, month, costType string) string {
```
Tested the changes, and they LGTM. Converting `row` into a struct is a good improvement.

I suggest adding a comment in the `createDetailedQuery` function to explicitly state that it's the same as  `createStandardQuery` with the addition of the `effective_price` column.

ishleenk17 commented on x-pack/metricbeat/module/gcp/billing/billing.go:
```diff
@@ -315,26 +380,134 @@ func generateEventID(currentDate string, rowItems []bigquery.Value) string {
+	IFNULL(SAFE_CAST(price.effective_price AS float64), 0) AS effective_price,
+	ARRAY_TO_STRING(ARRAY(
+	    SELECT
+	        CONCAT(t.key, ':', t.value)
```
This will result into strings something like this ?
"t1.key:t1.value" , "t2.key:t2.value", ......

gpop63 replied:
Yes, it would concatenate the keys and values resulting in the format you described above, `k1:v1,k2:v2` which makes it easy to translate into a map.

ishleenk17 commented on x-pack/metricbeat/module/gcp/billing/billing.go:
```diff
@@ -315,26 +380,134 @@ func generateEventID(currentDate string, rowItems []bigquery.Value) string {
+	service_id ASC,
+	service_description ASC,
+	effective_price ASC,
+	tags_string ASC;`,
```
Does it make sense to have order by on all the fields ?

ishleenk17 commented on x-pack/metricbeat/module/gcp/billing/billing_test.go:
```diff
@@ -5,11 +5,16 @@
 	"strconv"
 	"testing"
 
+	"cloud.google.com/go/bigquery"
```
bigquery was used before as well. How was it working without inclusion of this library ?

gpop63 replied:
It's because the function that translates the tags string into a map expects an argument of type `bigquery.Value`.

https://github.com/elastic/beats/blob/61e984ac49abb059fdd0198de38e979b8ab4e2eb/x-pack/metricbeat/module/gcp/billing/billing_test.go#L46

zmoog commented on x-pack/metricbeat/module/gcp/billing/billing.go:
```diff
@@ -266,34 +302,63 @@ func (m *MetricSet) queryBigQuery(ctx context.Context, client *bigquery.Client,
+	}
+
+	if isDetailedTable(tableName) {
+		_, _ = event.MetricSetFields.Put("effective_price", row.EffectivePrice)
```
Shouldn't we also add all the new fields to the mapping?

zmoog replied:
Oh, I guess this is already covered by the existing mapping:

```yaml
        - name: "metrics.*.*.*.*"
          type: object
          object_type: double
          object_type_mapping_type: "*"
          description: >
            Metrics that returned from Google Cloud API query.
```

right?

gpop63 replied:
Yeah, will add the fields to the mapping!

zmoog reviewed (commented):
The change LGTM, but we also need to update the docs.

We need to explain the "standard" vs. "detailed" tables and how to set up the metricset to get one or the other. 

We should highlight which extra fields you get from the detailed table. 

We should show a sample document and configuration for both standard and detailed cases at https://www.elastic.co/guide/en/beats/metricbeat/current/metricbeat-metricset-gcp-billing.html

zmoog reviewed (changes requested):
We should update the docs before merging the PR.

lucian-ioan commented on x-pack/metricbeat/module/gcp/billing/billing.go:
```diff
@@ -302,10 +367,10 @@ func getCurrentDate() string {
 	// create eventID using hash of current_date + invoice.month + project.id + project.name
 	// This will prevent more than one billing metric getting collected in the same day.
-	eventID := currentDate + rowItems[0].(string) + rowItems[1].(string) + rowItems[2].(string)
+	eventID := currentDate + row.InvoiceMonth + row.ProjectId + row.ProjectName
```
My concern with this implementation is that we no longer have the original event which aggregates `total_exact`. Instead, we now have all of the events that add up to `total_exact`.

There is also an issue where the ID of the event is always the same in `generateEventID` for all documents, resulting in only one document.

I think we should have the events grouped by sku/service as well as the original event that is currently used in dashboards. WDYT @zmoog?

zmoog replied:
@kaiyan-sheng, IIRC, you worked on this metricset in the past. I would love to read your comments when you have time.

@gpop63, do we also need to update the dashboard?

gpop63 replied:
@zmoog We'll have access to more granular data like `sku_id`, `service_id`, and `effective_price`, which will allow us to create different lenses for breaking down the data by these fields. However, given that we previously had only 1 document per day, the aggregation functions in current lenses must change from `max` to `sum`.

I'm also considering how to best manage the `tags` field so they might be used as a filter in the dashboard. The issue with using the `object` type is that the tags can't be leveraged for filtering since we don't know the mapping beforehand.

ishleenk17 replied:
> My concern with this implementation is that we no longer have the original event which aggregates `total_exact`. Instead, we now have all of the events that add up to `total_exact`.

I am not clear on this @lucian-ioan .
The total exact will be calculated per row after the grouping. And I think that is the right thing to do. 
Initially it was a sum of all the total_exact's we are now getting in different rows as ween [here](https://github.com/elastic/beats/pull/36902#issuecomment-1775621855).

zmoog replied:
@gpop63, in this PR, it is essential to ensure all the existing features continue to work while we add new fields. 

It's okay to add new visualizations later in a dedicated PR.

gpop63 commented:
The original implementation grouped the data by few fields, resulting in a single document. The updated implementation allows for a more granular view of the data, with each document containing detailed information about each service, SKU etc.

| Row | invoice_month | project_id | project_name | billing_account_id | cost_type | sku_id | sku_description | service_id | service_description | effective_price | tags_string | total_exact |
| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
| 1 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | CF4E-A0C7-E3BF | E2 Instance Core running in Americas | 6F81-5844-456A | Compute Engine | 0.02181159 | compute.googleapis.com/cores:2,compute.googleapis.com/machine_spec:e2-medium,compute.googleapis.com/memory:4096 | 3.216356 |  
| 2 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | F449-33EC-A5EF | E2 Instance Ram running in Americas | 6F81-5844-456A | Compute Engine | 0.00292353 | compute.googleapis.com/cores:2,compute.googleapis.com/machine_spec:e2-medium,compute.googleapis.com/memory:4096 | 1.724385 |  
| 3 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 6AE1-525F-8B80 | Balanced PD Capacity | 6F81-5844-456A | Compute Engine | 0.1 |   | 0.615075 |  
| 4 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 9DE9-9092-B3BC | Network Internet Egress from Americas to China | 6F81-5844-456A | Compute Engine | 0.23 |   | 0.005543 |  
| 5 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | DE9E-AFBC-A15A | Network Inter Zone Egress | 6F81-5844-456A | Compute Engine | 0.01 |   | 0.001505 |  
| 6 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | FDBC-6E3B-D4D8 | Network Internet Egress from Americas to Australia | 6F81-5844-456A | Compute Engine | 0.19 |   | 0.000561 |  
| 7 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | AED0-3315-7B11 | Networking Traffic Egress GCP Replication within Northern America | 95FF-2EF5-5EA1 | Cloud Storage | 0.02 |   | 0.000104 |  
| 8 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 0D5D-6E23-4250 | Standard Storage US Multi-region | 95FF-2EF5-5EA1 | Cloud Storage | 0.026 |   | 1.9e-05 |  
| 9 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 7C79-34BD-F61B | Network Inter Region Egress from Americas to Jakarta | 6F81-5844-456A | Compute Engine | 0.15 |   | 1.5e-05 |  
| 10 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 984A-1F27-2D1F | Network Egress via Carrier Peering Network - Americas Based | 6F81-5844-456A | Compute Engine | 0.04 |   | 1.3e-05 |  
| 11 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 9ADA-9AED-1B24 | Multi-Region Standard Class A Operations | 95FF-2EF5-5EA1 | Cloud Storage | 0.01 |   | 1e-05 |


| Row | invoice_month | project_id | project_name | billing_account_id | cost_type | total_exact |
|--|--|--|--|--|--|--|
| 1 | 202310 | robust-catalyst-399814 | My Project 70130 | 011702-58A742-BEB4E7 | regular | 5.563586 |

mrg-elastic committed:
update docs and fields

mrg-elastic committed:
add sku and service ids in hash

zmoog commented on x-pack/metricbeat/module/gcp/billing/_meta/docs.asciidoc:
```diff
@@ -1,18 +1,28 @@
+standard] and https://cloud.google.com/billing/docs/how-to/export-data-bigquery-tables/detailed-usage#detailed-usage-cost-data-schema[
+detailed] data schema documentation.
+
+For regular cost monitoring and analysis, standard tables utilize the
+`gcp_billing_export_v1_<BILLING_ACCOUNT_ID>` format, presenting a comprehensive
+view of Google Cloud expenses.
+
+Alternatively, detailed tables, identifiable by the `gcp_billing_export_resource_v1_<BILLING_ACCOUNT_ID>`
+format, offer richer data. They include additional fields, such as `effective_price`,
+enabling a more granular view of expenses.
+
 
```
We should explicitly state, "To get X, you must set Y". 

So "To get regular cost monitoring and analysis, you must set table pattern to `gcp_billing_export_v1_<BILLING_ACCOUNT_ID>` format", if I understand correctly.

mrg-elastic committed:
improve docs

ishleenk17 commented:
> The original implementation grouped the data by few fields, resulting in a single document every 24 hours. The updated implementation allows for a more granular view of the data, with each document containing detailed information about each service, SKU etc.
> 
> In total there are 70 rows, but I only copied the ones that have `total_exact` bigger than 0, meaning that there would be 70 documents created in elasticsearch.
> 
> Updated implementation
> Current implementation
> Row	invoice_month	project_id	project_name	billing_account_id	cost_type	total_exact
> 1	202310	robust-catalyst-399814	My Project 70130	011702-58A742-BEB4E7	regular	5.563586

@gpop63 : This is because we are now also grouping by sku.id, service.id etc fields in the query?

gpop63 commented:
> @gpop63 : This is because we are now also grouping by sku.id, service.id etc fields in the query?

@ishleenk17 Yes, adding more fields to the `GROUP BY` clause, like `sku.id` and `service.id`, increases the granularity of the grouping. This creates more unique combinations to group by, hence resulting in a higher number of rows in the query results.

ishleenk17 commented:
> > @gpop63 : This is because we are now also grouping by sku.id, service.id etc fields in the query?
> 
> @ishleenk17 Yes, adding more fields to the `GROUP BY` clause, like `sku.id` and `service.id`, increases the granularity of the grouping. This creates more unique combinations to group by, hence resulting in a higher number of rows in the query results.

Thanks, I hope this is inline with the grouping activity going on for the TSDB front.

ishleenk17 commented:
Shall we create a separate ticket for changes to the dashboards?
Some additions to the dashboards: 

- lens visualisations where a customer can view cost/ sku or cost/service
- Filtering by tags?
- Effective Price view

Also, some revamping would be needed to the existing dashboard.
Maybe as part of this PR we can ensure that nothing breaks in the current dashboard.

And add the other visualisations to the dashboard as part of another PR. 

@zmoog : Thoughts ?

gpop63 commented on x-pack/metricbeat/module/gcp/billing/_meta/fields.yml:
```diff
@@ -15,3 +15,22 @@
+  - name: service_description
+    type: text
+    description: The Google Cloud service that reported the Cloud Billing data.
+  - name: tags
+    type: object
+    object_type: keyword
+    description: A collection of key-value pairs that provide additional metadata.
```
One small problem I see is if we try to filter the dashboard by tags. We probably can't use the `object` type because we don't know the fields beforehand. What if we use something like `nested` instead? That way, the field definitions would look like this:

```yaml
- name: tags
  type: nested
  fields:
  - name: key
    type: keyword
  - name: value
    type: keyword
```

I'm wondering if this is a good option so the dashboard can be filtered by tags keys and values. If yes, then the `createTagsMap` function must be changed because `nested` types expect a JSON array.

ishleenk17 replied:
IIIRC, Rajvi had tried out visualizations with object fields.
cc: @rajvi-elastic

rajvi-elastic replied:
Yes @ishleenk17, Filters can be applied on both object and nested fields. It won't throw any errors.

gpop63 replied:
What I'm trying to highlight is that with the current implementation where `tags` is a map, we can't create input control groups to filter dashboard data using tags keys or tags values fields.

<b>Current implementation:</b>

```json
{
  "gcp": {
    "billing": {
      ...
      "tags": {
        "k1": "v1",
        "k2": "v2"
      }
    }
  }
}
```

<b>Storing tags as an array of nested tag objects:</b>

```json
{
  "gcp": {
    "billing": {
      ...
      "tags": [
        {
          "value": "v1",
          "key": "k1"
        },
        {
          "value": "v2",
          "key": "k2"
        }
      ]
    }
  }
}
```

The problem with using a map is that in an input control group you can't use `tags.key` or `tags.value` as a field because the mapping is "dynamic" and you can't know the tags beforehand.

![gcp_billing_control](https://github.com/elastic/beats/assets/94497545/accac0dd-4b62-46db-8a0b-a26000a206d0)

Using `nested` type allows us to filter by `gcp.billing.tags.key` or `gcp.billing.tags.value`.

![gcp_billing_control_updated](https://github.com/elastic/beats/assets/94497545/20995046-bf95-4def-a5f3-05fcb591f2ea)

I'm unsure if filtering by `key` and `value` separately is ideal, or if it should be done by the combination of `key+value`.

ishleenk17 replied:
As per me a concat of key+value makes more sense for filtering, rather than individual filtering on just key or just value. 
@kaiyan-sheng @zmoog any thoughts on this since you both have worked on billing modules of other cloud packages.

ishleenk17 replied:
Lets have the key and value as separate fields.
If need be we can have the key+value as a concat in the pipeline. and use that for filtering in dashboards.

zmoog commented:
> Shall we create a separate ticket for changes to the dashboards?

Yeah, I would make sure the existing dashboard continues to work as expected. We can leverage the new fields in a different PR.

Commit:
Merge branch 'main' into add_gcp-billing_fields

ishleenk17 commented:
@gpop63 : Lets create a separate ticket for creating updated dashbaords with the new change. 
Can you please point out what is not working in the current dashboard with the new changes.

Commit:
Merge branch 'main' into add_gcp-billing_fields

mrg-elastic committed:
update fields types

mrg-elastic committed:
use array instead of map for tags
include tags in the event signature

ishleenk17 commented:
@gpop63 : Changing the nested to array will bring a change in the geenrated event as well. Lets update that as well.
Also, please update the dashboard

mrg-elastic committed:
update sample doc

mrg-elastic committed:
use sum instead of max for dashboards

zmoog commented:
@gpop63, why are we changing the tags structure and the aggregation type for the `gcp.billing.total` field (isn't this one of the preexisting fields)?

zmoog commented on x-pack/metricbeat/module/gcp/_meta/kibana/7/lens/057de170-e88d-11ea-bf8c-d13ebf358a78.json:
```diff
@@ -14,7 +14,7 @@
                                     "dataType": "number",
                                     "isBucketed": false,
                                     "label": "Maximum of gcp.billing.total",
-                                    "operationType": "max",
+                                    "operationType": "sum",
```
If we change the aggregation type, we should also update the label accordingly (now it's named "Maximum of [..]").

But I am mostly interested in the change of aggregation type.

gpop63 replied:
Oh yeah I forgot about the label. Previously, with only one document per day, we used `max` to display the cost. However, since the data granularity has increased, generating more documents daily, we need to use `sum` to accurately aggregate the costs from all documents.

zmoog replied:
So now we have multiple documents per day, instead of just one. Got it.

We should then update the PR description to reflect this change. The PR description should contain a TL;DR summary of all the changes happening.

gpop63 commented:
@zmoog If tags field is an object, adding an input control group in the dashboard for filtering based on tags keys and/or values is not possible due to the dynamic mapping.

The field definition below allows us to filter by tags keys and/or values because there is a mapping which allows us to choose from either `gcp.billing.tags.key` or `gcp.billing.tags.value` in the input control group. I provided more details [here](https://github.com/elastic/beats/pull/36902#discussion_r1374444469).

```yaml
- name: tags
  type: nested
  fields:
  - name: key
    type: keyword
  - name: value
    type: keyword
```

zmoog commented:
> @zmoog If tags field is an object, adding an input control group in the dashboard for filtering based on tags keys and/or values is not possible due to the dynamic mapping.
       
> The field definition below enables filtering by tags keys and/or values because there is a mapping which allows us to choose from either `gcp.billing.tags.key` or `gcp.billing.tags.value` in the input control group. I provided more details [here](https://github.com/elastic/beats/pull/36902#discussion_r1374613815).

Thanks for the clarification.

zmoog commented:
@gpop63, is this PR going to make https://github.com/elastic/integrations/pull/5665 obsolete?

ishleenk17 reviewed (approved):
Lets make the PR slight more descriptive regarding the changes in the PR.
Otherwise, looks good!

gpop63 commented:
> @gpop63, is this PR going to impact [elastic/integrations#5665](https://github.com/elastic/integrations/pull/5665) or make it obsolete?

Yeah, kind of. This [PR](https://github.com/elastic/integrations/pull/7249/files) probably made that one obsolete. Also the aggregation change to `last_value` will not work since we have to use `sum` after this change.

mrg-elastic committed:
fix label
minor changes

gpop63 commented:
/test
